clippy_lints/src/shadow_type.rs

```diff
@@ -0,0 +1,105 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::{GenericParamKind, Generics, HirId, Item, ItemKind, Mod};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_session::impl_lint_pass;
+use rustc_span::{Ident, Span, Symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for generic parameters that shadow types in scope.
+    ///
+    /// ### Why restrict this?
+    /// To avoid confusion and potential bugs, as it can lead to
+    /// misunderstandings about which type is being referred to.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// struct Foo;
+    /// struct Bar<Foo> { f: Foo }
+    /// ```
+    ///
+    /// Use instead:
+    /// ```no_run
+    /// struct Foo;
+    /// struct Bar<F> { f: F } // use different generic parameter name
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub SHADOW_TYPE_GENERIC,
+    restriction,
+    "shadowing of type in scope by generic parameter"
+}
+
+#[derive(Default)]
+pub(crate) struct ShadowTypeGeneric {
+    module_types: FxHashMap<Symbol, Span>,
+}
+
+impl_lint_pass!(ShadowTypeGeneric => [SHADOW_TYPE_GENERIC]);
+
+impl<'tcx> LateLintPass<'tcx> for ShadowTypeGeneric {
+    fn check_mod(&mut self, cx: &LateContext<'tcx>, module: &'tcx Mod<'tcx>, _: HirId) {
+        self.module_types.clear();
+        let items = module.item_ids.iter().map(|&id| cx.tcx.hir_item(id));
+        for item in items {
+            if item.span.in_external_macro(cx.sess().source_map()) || item.span.from_expansion() {
```

Comment:
`item.span.from_expansion()` covers any case where `in_external_macro` would be true